pkg/services/provisioning/dashboards/file_reader.go

```diff
@@ -98,25 +104,40 @@ func (fr *FileReader) walkDisk() error {
 
 	fr.handleMissingDashboardFiles(provisionedDashboardRefs, filesFoundOnDisk)
 
-	sanityChecker := newProvisioningSanityChecker(fr.Cfg.Name)
-
+	usageTracker := newUsageTracker()
 	if fr.FoldersFromFilesStructure {
-		err = fr.storeDashboardsInFoldersFromFileStructure(filesFoundOnDisk, provisionedDashboardRefs, resolvedPath, &sanityChecker)
+		err = fr.storeDashboardsInFoldersFromFileStructure(filesFoundOnDisk, provisionedDashboardRefs, resolvedPath, usageTracker)
 	} else {
-		err = fr.storeDashboardsInFolder(filesFoundOnDisk, provisionedDashboardRefs, &sanityChecker)
+		err = fr.storeDashboardsInFolder(filesFoundOnDisk, provisionedDashboardRefs, usageTracker)
 	}
 	if err != nil {
 		return err
 	}
 
-	sanityChecker.logWarnings(fr.log)
+	fr.mux.Lock()
+	defer fr.mux.Unlock()
 
+	fr.usageTracker = usageTracker
```

Comment:
This is a kind of implementation of a non-blocking algorithm. I thought that resetting a tracker will cost time and block the entire process.

About concurrent access - yes, it can be accessed concurrently. Duplicates validator checks that there are no duplicates across all readers, and it requires access to usage trackers of the readers.
Corresponding link [here](https://github.com/grafana/grafana/pull/26742/files/1f0a9f0d40154eb7b3d1443b12be53f5478bebc8#diff-40968223800ec29c8675bd383f59f205f60d58c17284eb2abd72f5d4e916333aR65-R66).

Reply:
That link doesn't work any longer since you force pushed. 

The `getUsageTracker()` does a `RLock` followed by a `RUnlock`, but I do notice you do a `Lock` followed by `Unlock` above the mention code above so please disregard my comment.